### PR TITLE
Revert 'Added jackson dependency to server" and change extension reading

### PR DESCRIPTION
**Pull request by ryanbogan**
Signed-off-by: Ryan Bogan <rbogan@amazon.com>

### Description
Reverts https://github.com/opensearch-project/OpenSearch/pull/5366.  Changes extension reading to use SnakeYAML instead of jackson-databind

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/5504

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java [hunk 1]
```diff
@@ -556,10 +555,35 @@ public String executor() {
     }
 
     private ExtensionsSettings readFromExtensionsYml(Path filePath) throws IOException {
-        ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory());
-        InputStream input = Files.newInputStream(filePath);
-        ExtensionsSettings extensionSettings = objectMapper.readValue(input, ExtensionsSettings.class);
-        return extensionSettings;
+        Yaml yaml = new Yaml();
+        InputStream inputStream = Files.newInputStream(filePath);
```
Comment (uschindler) [inline server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:559]: Please use try with resources. This may produce file handle leaks.

**Review by uschindler — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java (hunk 1 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:559]: Should the try block include just those lines or more?  The only place the method is called currently has the method in a try block

Comment (uschindler) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:559]: No like that:

```java
    try (InputStream inputStream = Files.newInputStream(filePath)) {
        Map<String, Object> obj = yaml.load(inputStream);
        if (obj == null) {
            throw new IOException("extensions.yml is empty");
        }
        List<HashMap<String, ?>> unreadExtensions = new ArrayList<>((Collection<HashMap<String, ?>>) obj.get("extensions"));
        List<Extension> readExtensions = new ArrayList<Extension>();
        for (Map<String, ?> extensionMap : unreadExtensions) {
            readExtensions.add(
                new Extension(
                    extensionMap.get("name").toString(),
                    extensionMap.get("uniqueId").toString(),
                    extensionMap.get("hostName").toString(),
                    extensionMap.get("hostAddress").toString(),
                    extensionMap.get("port").toString(),
                    extensionMap.get("version").toString(),
                    extensionMap.get("description").toString(),
                    extensionMap.get("opensearchVersion").toString(),
                    extensionMap.get("javaVersion").toString(),
                    extensionMap.get("className").toString(),
                    extensionMap.get("customFolderName").toString(),
                    extensionMap.get("hasNativeController").toString()
                )
            );
        }
    }
```

Comment (uschindler) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:559]: FYI I was talking about try-with-resources that ensures that files and other closeable resources are correctly closed in all error conditions.

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:559]: Oh that makes more sense

**Review by peternied — APPROVED** [review]
I like it @ryanbogan, much cleaner

File: server/src/test/java/org/opensearch/common/settings/WriteableSettingTests.java [hunk 2]
```diff
@@ -462,6 +463,7 @@ public void testVersionSetting() throws IOException {
         }
     }
 
+    @Ignore
```
Comment (uschindler) [inline server/src/test/java/org/opensearch/common/settings/WriteableSettingTests.java:466]: You can fix this test in the following way: Add a setter marked as package-private (no access flags) in the s
Settings class to modify the default. By that only a class from the same package can modify the default value, not code anywjere outside.

WARNING: If the fields to modify are final, don't test this at all, kill this test!

**Review by uschindler — CHANGES_REQUESTED** [review]
Otherwise looks good! Thanks!

**Comment by ryanbogan** [comment]
Failing test is flaky: https://github.com/opensearch-project/OpenSearch/issues/5766

**Comment by peternied** [comment]
> Reverts https://github.com/opensearch-project/OpenSearch/pull/5366.

I disagree with remove jackson-databind - pragmatically this is a considerable burden on all the plugin times migrated to get Jackson from OpenSearch.  Regardless of our feelings of the quality of jackson-databind, it is still popular in the OpenSearch project, complete removal will require effort on the verge of a release.

I suggest that **all** jackson dependencies should be transitively included by OpenSearch for security patching purposes.

**Review by peternied — CHANGES_REQUESTED** [review]
I thought there was a way to keep the dependency in without elevated the permissions in server, is this not the case?

**Comment by uschindler** [comment]
> > Reverts #5366.
> 
> I disagree with remove jackson-databind - pragmatically this is a considerable burden on all the plugin times migrated to get Jackson from OpenSearch. Regardless of our feelings of the quality of jackson-databind, it is still popular in the OpenSearch project, complete removal will require effort on the verge of a release.
> 
> I suggest that **all** jackson dependencies should be transitively included by OpenSearch for security patching purposes.

For that we have BOM maven dependencies. You can have an Opensearch BOM that have fixed versions for jackson, but jackson-databind should never ever be in server's classpath.

The problem here is: whe each plugin has its own JAR file in its private classpath (shiedled by classloader), there are no malicious interactions between the instances as each only sees its classloader. Without sandboxing, jackson-databind can easily be used to create malicious code if ask it to deserialize classes, also those in server core. So, if jackson-databind is in server's classpath there is a chance that a malicious plugin may deserialize and trigger updates in fields of classes values in server's classpath.

If the plugin has its own copy of jackson databind, due to shielding it is not possible to update core's classes, especially when OpenSearch has moved to module system (Elasticsearch aleady did this).

**Comment by uschindler** [comment]
So please don't do this. Use a POM of opensearch to force versions for downstream code.

**Review by peternied — APPROVED** [review]
>>I thought there was a way to keep the dependency in without elevated the permissions in server, is this not the case?

>Yes it works, but a plugin can still trick jackson-databind to access server's classes if it is not only in its private module/classloader.

Well damn.  Thanks for the details.  Lets get this in

**Review by saratvemulapalli — APPROVED** [review]

**Comment by uschindler** [comment]
Hi,

> > > I thought there was a way to keep the dependency in without elevated the permissions in server, is this not the case?
> 
> > Yes it works, but a plugin can still trick jackson-databind to access server's classes if it is not only in its private module/classloader.
> 
> Well damn. Thanks for the details. Lets get this in

I have to admit that without using the Java Module System, it is still possible that a plugin may change server's data using deep reflection, as only the module system prevents setAccessible() when the core does not "open" the module to specific submodules. So a stong deep reflection protection is only there if the Java moudle system is used (this was already suggested by @rmuir).

But the revert is also useful to actually migrate to the module system, because if you don't use a dependency it should never ever be part of the "requires" clauses of a module. Sure, you could exclude it in the `module-info.java` and still keep it in dependencies, but then downstream module users won't see the classes anyways, so adding it as a mvane/gradle dependency would not bring anything.

So my proposal to proceed would be:
- Adapt Java Module System (JMS) - Elasticsearch already did this, gladly they had help by one of the OpenJDK committers hired by them
- Clean up dependency hell and also remove any JAR files from dependencies with are neither useable as real module or auto-module. In fact you have to decide here: Write a bit more code or add dependency hell entries. Sorry, jackson-databind is one of the dependencies to remove on longer term, although it is used often in open source. But somebody has to stop such outdated dependecies and move to "module system compatible ones" using no deep reflection and instead use `MethodHandle.Lookup` instances that were instantiated by the consumer code and passed to serialization lib to give it access to private APIs.
- In case of dependencies that cannot be included as real modules, shade them into another JAR file and make a module together with the plugin. This may be a good way to still use jackson-databind.

**Comment by dblock** [comment]
@uschindler As an additional option to JMS, a handful of folks (w/@saratvemulapalli) are working on extracting an SDK out of OpenSearch (https://github.com/opensearch-project/opensearch-sdk-java) and stop plugins from taking direct dependencies on OpenSearch. Ultimately that should allow full sandboxing of plugins (extensions), including by containerizing them. Lots of words in https://github.com/opensearch-project/OpenSearch/issues/2447.

**Pull request merged**